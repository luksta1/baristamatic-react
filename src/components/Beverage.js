import React from 'react';
import { Card, Icon, Image } from 'semantic-ui-react'


const Beverage = (props) => {

    // logic to check if the beverage is in stock or not and set the tag line appropriately
    let stock = '';
    if (props.beverage.inStock) {
        stock += 'Available'
    } else {
        stock += 'Not Available'
    }
    // function to run an order once a drink is click on. future refactoring would include dimming out a drink once it becomes unavailable
    const runOrder = () => {
        props.calculateStock(props.bev)
    }

    return (
        <Card onClick={runOrder} id={props.bev} className="beverageCard">
            <Image src={props.beverage.image} />
            <Card.Content>
                <Card.Header>
                    {props.beverage.name}
                </Card.Header>
                <Card.Meta>
                    <span className='price'>
                        ${props.beverage.price.toFixed(2)}
                    </span>
                </Card.Meta>
            </Card.Content>
            <Card.Content extra>
                <Icon name='info circle' />
                {stock}

            </Card.Content>
        </Card>

    )
}

export default Beverage