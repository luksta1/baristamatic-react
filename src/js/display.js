const mapInventory = (inv) => {
    console.log('Inventory:')
    Object.keys(inv)
        .sort()
        .forEach(ingredient => {
            console.log(`${inv[ingredient].name}, ${inv[ingredient].stock}`)
        });
}

const mapMenu = (menu) => {
    console.log('Menu:')
    const beverages = Object.keys(menu);
    beverages.sort()
        .forEach(beverage => {
            if (menu.hasOwnProperty(beverage)) {
                const index = beverages.indexOf(beverage)
                console.log(`${index + 1}, ${menu[beverage].name}, $${menu[beverage].price.toFixed(2)}, ${menu[beverage].inStock}`)
            }
        });
}

module.exports = {
    mapInventory,
    mapMenu
}
