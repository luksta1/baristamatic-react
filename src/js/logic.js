const readline = require('readline');
const { mapInventory, mapMenu } = require('./display');

const inStockCheck = (menu, inv) => {
    for (let beverage in menu) {
        if (menu.hasOwnProperty(beverage)) {
            for (let ingredient in menu[beverage].ingredients) {
                if (menu[beverage].ingredients.hasOwnProperty(ingredient)) {
                    if (menu[beverage].ingredients[ingredient] <= inv[ingredient].stock) {
                        menu[beverage].inStock = true
                    } else {
                        menu[beverage].inStock = false
                        break;
                    }
                }
            }
        }
    }
}

const restock = (menu, inv) => {
    for (let ingredient in inv) {
        if (inv.hasOwnProperty(ingredient)) {
            inv[ingredient].stock = 10
        }
    }
    inStockCheck(menu, inv)
}

const calculateStock = (beverage, inv, menu) => {
    for (let ingredient in menu[beverage].ingredients) {
        if (menu[beverage].ingredients.hasOwnProperty(ingredient)) {
            inv[ingredient].stock = inv[ingredient].stock - menu[beverage].ingredients[ingredient]
        }
    }
    inStockCheck(menu, inv)
}

const runOrder = (input, menu, inv) => {
    const menuItems = Object.keys(menu).sort();
    let selection = Number(input);

    if (!isNaN(selection) && selection > 0 && selection <= menuItems.length) {
        let bevOrdered = menu[menuItems[selection - 1]]
        if (!bevOrdered.inStock) {
            console.log(`Out Of Stock: ${bevOrdered.name}`)
        } else {
            console.log(`Dispensing: ${bevOrdered.name}`)
            calculateStock(menuItems[selection - 1], inv, menu);
            mapInventory(inv);
            mapMenu(menu)
        }
    } else {
        console.log(`Invalid Selection: ${input}`)
    }

}

const runCoffeeMachine = (inv, menu) => {

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });

    rl.on('line', (input) => {
        switch (input.trim()) {
            case 'r':
            case 'R':
                console.log('Restocking Machine');
                restock(menu, inv)
                mapInventory(inv);
                mapMenu(menu)
                break;
            case 'q':
            case 'Q':
                process.exit(0);
                break;
            default:
                runOrder(input, menu, inv)
                break;
        }
        rl.prompt();
    }).on('close', () => {
        process.exit(0);
    });
}

module.exports = {
    restock,
    inStockCheck,
    calculateStock,
    runCoffeeMachine
}
