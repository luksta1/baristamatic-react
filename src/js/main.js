const { menu, inventory } = require('./menu-inventory')
const { mapInventory, mapMenu } = require('./display');
const { restock, inStockCheck, runCoffeeMachine } = require('./logic');

const main = () => {
    restock(menu, inventory)
    inStockCheck(menu, inventory)
    mapInventory(inventory);
    mapMenu(menu)
    runCoffeeMachine(inventory, menu)
}

main();

