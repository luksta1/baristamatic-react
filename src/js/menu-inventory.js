let inventory = {
    coffee: {
        name: 'Coffee',
        stock: 10,
        unitPrice: 0.75
    },
    decafCoffee: {
        name: 'Decaf Coffee',
        stock: 10,
        unitPrice: 0.75
    },
    sugar: {
        name: 'Sugar',
        stock: 10,
        unitPrice: 0.25
    },
    cream: {
        name: 'Cream',
        stock: 10,
        unitPrice: 0.25
    },
    steamedMilk: {
        name: 'Steamed Milk',
        stock: 10,
        unitPrice: 0.35
    },
    foamedMilk: {
        name: 'Foamed Milk',
        stock: 10,
        unitPrice: 0.35
    },
    espresso: {
        name: 'Espresso',
        stock: 10,
        unitPrice: 1.10
    },
    cocoa: {
        name: 'Cocoa',
        stock: 10,
        unitPrice: 0.90
    },
    whippedCream: {
        name: 'Whipped Cream',
        stock: 10,
        unitPrice: 1.00
    }
}

let menu = {
    coffee: {
        name: 'Coffee',
        ingredients: {
            coffee: 3,
            sugar: 1,
            cream: 1,
        },
        price: 2.75,
        inStock: true
    },
    decafCoffee: {
        name: 'Decaf Coffee',
        ingredients: {
            decafCoffee: 3,
            sugar: 1,
            cream: 1
        },
        price: 2.75,
        inStock: true
    },
    caffeLate: {
        name: 'Caffe Late',
        ingredients: {
            espresso: 2,
            steamedMilk: 1
        },
        price: 2.55,
        inStock: true
    },
    caffeAmericano: {
        name: 'Caffe Americano',
        ingredients: {
            espresso: 3
        },
        price: 3.30,
        inStock: true
    },
    caffeMocha: {
        name: 'Caffe Mocha',
        ingredients: {
            espresso: 1,
            cocoa: 1,
            steamedMilk: 1,
            whippedCream: 1
        },
        price: 2.70,
        inStock: true
    },
    cappuccino: {
        name: 'Cappuccino',
        ingredients: {
            espresso: 2,
            steamedMilk: 1,
            foamedMilk: 1
        },
        price: 2.90,
        inStock: true
    }
}

module.exports = {
    inventory,
    menu
}
