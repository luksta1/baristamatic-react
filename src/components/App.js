import React, { Component } from 'react';
import Beverage from './Beverage.js'
import { menu, inventory } from '../js/menu-inventory.js'
import InventoryComp from './InventoryComp.js';
import Header from './Header.js';

class App extends Component {

  constructor(props) {
    super(props);
    this.state = {
      menu: { ...menu },
      inventory: { ...inventory }
    }
    this.restock = this.restock.bind(this)
    this.inStockCheck = this.inStockCheck.bind(this)
    this.calculateStock = this.calculateStock.bind(this)
  }


  // run function to check inventory stock each time the app is loaded
  componentWillMount = () => {
    this.inStockCheck();
  }

  // function to reset the inventory on state appropriately
  inStockCheck = () => {
    let s = this.state
    let newState = { ...this.state }
    for (let beverage in s.menu) {
      for (let ingredient in s.menu[beverage].ingredients) {
        if (s.menu[beverage].ingredients[ingredient] <= s.inventory[ingredient].stock) {
          newState.menu[beverage].inStock = true
        } else {
          newState.menu[beverage].inStock = false
          break;
        }
      }
    }
    this.setState({ newState })
  }

  // function to reset the stock on state appropriately, which then runs the function to check stock
  restock = () => {
    for (let ingredient in this.state.inventory) {
      if (this.state.inventory.hasOwnProperty(ingredient)) {
        let newState = { ...this.state }
        newState.inventory[ingredient].stock = 10
        this.setState({ newState })
      }
    }
    this.inStockCheck()
  }

  // function to calculate the stock on state based on an ordered drink, when then runs the function to check stock
  calculateStock = (beverage) => {
    let newState = { ...this.state }
    for (let ingredient in this.state.menu[beverage].ingredients) {
      if (this.state.menu[beverage].ingredients.hasOwnProperty(ingredient)) {
        newState.inventory[ingredient].stock = this.state.inventory[ingredient].stock - this.state.menu[beverage].ingredients[ingredient]
      }
      if (newState.inventory[ingredient].stock < 0) {
        newState.inventory[ingredient].stock = 0
      }
    }
    this.setState({ newState })
    this.inStockCheck()
  }

  // render of the main app screen, header, inventory and mapping over the beverages
  render() {
    const menuItems = Object.keys(this.state.menu).sort();
    return (
      <div className="App">
        <div className="head">
          <Header restock={this.restock.bind(this)} />
        </div>
        <div className="main">
          <div className="itemInventory">
            <InventoryComp inventory={this.state.inventory} />
          </div>
          <div className="drinkMenu">
            {
              menuItems.map(beverage => {
                let bev = this.state.menu[beverage]
                return (
                  <Beverage calculateStock={this.calculateStock.bind(this)} bev={beverage} key={bev.name} beverage={bev} />
                )
              })}
          </div>
        </div>
      </div>
    );
  }
}

export default App;
