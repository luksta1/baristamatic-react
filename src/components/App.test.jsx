import App from './App';
import Beverage from './Beverage'
import { menu, inventory } from '../js/menu-inventory.js'

it('renders correctly', () => {
    const wrapper = shallow(
        <App />
    )

    expect(wrapper).toMatchSnapshot();
})

it('renders the menu on state correctly', () => {
    const wrapper = shallow(
        <App />
    )

    expect(JSON.stringify(wrapper.state().menu)).toBe(JSON.stringify({ ...menu }))
})

it('renders the inventory on state correctly', () => {
    const wrapper = shallow(
        <App />
    )

    expect(JSON.stringify(wrapper.state().inventory)).toBe(JSON.stringify({ ...inventory }))
})

it('renders the beverage component 6 times', () => {
    const wrapper = shallow(
        <App />
    )

    expect(wrapper.find(Beverage).length).toEqual(6);
})