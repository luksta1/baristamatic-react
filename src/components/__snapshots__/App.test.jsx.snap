// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <div
    className="head"
  >
    <Header
      restock={[Function]}
    />
  </div>
  <div
    className="main"
  >
    <div
      className="itemInventory"
    >
      <InventoryComp
        inventory={
          Object {
            "cocoa": Object {
              "name": "Cocoa",
              "stock": 10,
              "unitPrice": 0.9,
            },
            "coffee": Object {
              "name": "Coffee",
              "stock": 10,
              "unitPrice": 0.75,
            },
            "cream": Object {
              "name": "Cream",
              "stock": 10,
              "unitPrice": 0.25,
            },
            "decafCoffee": Object {
              "name": "Decaf Coffee",
              "stock": 10,
              "unitPrice": 0.75,
            },
            "espresso": Object {
              "name": "Espresso",
              "stock": 10,
              "unitPrice": 1.1,
            },
            "foamedMilk": Object {
              "name": "Foamed Milk",
              "stock": 10,
              "unitPrice": 0.35,
            },
            "steamedMilk": Object {
              "name": "Steamed Milk",
              "stock": 10,
              "unitPrice": 0.35,
            },
            "sugar": Object {
              "name": "Sugar",
              "stock": 10,
              "unitPrice": 0.25,
            },
            "whippedCream": Object {
              "name": "Whipped Cream",
              "stock": 10,
              "unitPrice": 1,
            },
          }
        }
      />
    </div>
    <div
      className="drinkMenu"
    >
      <Beverage
        bev="caffeAmericano"
        beverage={
          Object {
            "image": "/assets/1_cafeamericano.png",
            "inStock": true,
            "ingredients": Object {
              "espresso": 3,
            },
            "name": "Caffe Americano",
            "price": 3.3,
          }
        }
        calculateStock={[Function]}
        key="Caffe Americano"
      />
      <Beverage
        bev="caffeLatte"
        beverage={
          Object {
            "image": "/assets/2_cafelatte.png",
            "inStock": true,
            "ingredients": Object {
              "espresso": 2,
              "steamedMilk": 1,
            },
            "name": "Caffe Latte",
            "price": 2.55,
          }
        }
        calculateStock={[Function]}
        key="Caffe Latte"
      />
      <Beverage
        bev="caffeMocha"
        beverage={
          Object {
            "image": "/assets/3_cafemocha.png",
            "inStock": true,
            "ingredients": Object {
              "cocoa": 1,
              "espresso": 1,
              "steamedMilk": 1,
              "whippedCream": 1,
            },
            "name": "Caffe Mocha",
            "price": 2.7,
          }
        }
        calculateStock={[Function]}
        key="Caffe Mocha"
      />
      <Beverage
        bev="cappuccino"
        beverage={
          Object {
            "image": "/assets/4_cappuccino.png",
            "inStock": true,
            "ingredients": Object {
              "espresso": 2,
              "foamedMilk": 1,
              "steamedMilk": 1,
            },
            "name": "Cappuccino",
            "price": 2.9,
          }
        }
        calculateStock={[Function]}
        key="Cappuccino"
      />
      <Beverage
        bev="coffee"
        beverage={
          Object {
            "image": "/assets/5_coffee.png",
            "inStock": true,
            "ingredients": Object {
              "coffee": 3,
              "cream": 1,
              "sugar": 1,
            },
            "name": "Coffee",
            "price": 2.75,
          }
        }
        calculateStock={[Function]}
        key="Coffee"
      />
      <Beverage
        bev="decafCoffee"
        beverage={
          Object {
            "image": "/assets/6_decaf.png",
            "inStock": true,
            "ingredients": Object {
              "cream": 1,
              "decafCoffee": 3,
              "sugar": 1,
            },
            "name": "Decaf Coffee",
            "price": 2.75,
          }
        }
        calculateStock={[Function]}
        key="Decaf Coffee"
      />
    </div>
  </div>
</div>
`;
